src/tools/Avalonia.Generators/NameGenerator/AvaloniaNameIncrementalGenerator.cs

```diff
@@ -0,0 +1,204 @@
+
+        // Actual parsing step. We input XAML files one by one, but don't resolve any types.
+        // That's why we use NoOp type system here, allowing parsing to run detached from C# compilation.
+        // Otherwise we would need to re-parse XAML on any C# file changed.
+        var parsedXamlClasses = xamlFiles
+            .Select(static (file, cancellationToken) =>
+            {
+                var text = file.GetText(cancellationToken);
+                var diagnostics = new List<DiagnosticDescriptor>();
+                if (text is not null)
+                {
+                    try
+                    {
+                        var xaml = text.ToString();
+                        var viewResolver = new XamlXViewResolver(s_noopCompiler);
+                        var view = viewResolver.ResolveView(xaml);
+                        if (view is null)
+                        {
+                            return null;
+                        }
+
+                        var nameResolver = new XamlXNameResolver();
+                        var xmlNames = nameResolver.ResolveXmlNames(view.Xaml);
+
+                        return new XmlClassInfo(
+                            new ResolvedXmlView(view, xmlNames),
+                            diagnostics.ToImmutableArray());
+                    }
+                    catch (Exception ex)
+                    {
+                        diagnostics.Add(GeneratorExtensions.NameGeneratorUnhandledError(ex));
+                        return new XmlClassInfo(null, diagnostics.ToImmutableArray());
+                    }
+                }
+
+                return null;
+            })
+            .Where(request => request is not null)
+            .WithTrackingName(TrackingNames.ParsedXamlClasses);
+
+        // IMPORTANT: we shouldn't cache CompilationProvider as a whole,
+        // But we also should keep in mind that CompilationProvider can frequently re-trigger generator.
+        var compiler = context.CompilationProvider
+            .Select(static (compilation, _) =>
+            {
+                var roslynTypeSystem = new RoslynTypeSystem(compilation);
+                return MiniCompiler.CreateRoslyn(roslynTypeSystem, MiniCompiler.AvaloniaXmlnsDefinitionAttribute);
+            })
+            .WithTrackingName(TrackingNames.XamlTypeSystem);
+
+        // Note: this step will be re-executed on any C# file changes.
+        // As much as possible heavy tasks should be moved outside of this step, like XAML parsing.
+        var resolvedNames = parsedXamlClasses
+            .Combine(compiler)
+            .Select(static (pair, _) =>
+            {
+                var (classInfo, compiler) = pair;
+                var hasDevToolsReference = compiler.TypeSystem.FindAssembly("Avalonia.Diagnostics") is not null;
+                var nameResolver = new XamlXNameResolver();
+
+                var diagnostics =  new List<DiagnosticDescriptor>(classInfo!.Diagnostics);
```

Comment:
Nit: consider using `ImmutableArray.CreateBuilder()` here and `diagnostics.MoveToImmutable()` to create the final array.